README.md
# typeit.js
Simple JavaScript Typing Animation Plugin

## Setup development environment with docker
Usage purposes: You don't (or you do not want to) have node installed globally on your machine.

Required:
- [docker](https://docs.docker.com/install/)
- [docker-compose](https://docs.docker.com/compose/)

Run `$ docker-compose up` inside the project's root directory.

## Usage
### Initialize

To initialize Typeit plugin you need to pass two arguments:
1. st - selector or DOM element
2. nd - configuration object

Init example:
```
<h1>
  <span id="typeit"></span>
</h1>

<script src="./path/to/typeit.js"></script>
<script>
new CCS.TypeIt('#typeit', {
  words: [
    'Sample text',
    ['This', 'is', 'also', 'allowed']
  ],
  speed: 100,
  delay: 900,
  backwards: true,
  infinity: true,
  onType: function(event, word) {
    // handle event
  }
})
</script>
```
### Options (config)
- `words` - an array of Strings (array with pieces of words will be transformed to strings). **Note**: text content typed inside your DOM element will be automatically added to the beginning of the words array
- `speed` - typing speed, by default `95`(ms)
- `delay` - delay between *typeit* actions, by default `650`(ms)
- `letterTag` - specify each letter html tag, by default `false` - pure text
- `letterClass` - specify each letter wrapper class, by default `false` - no class. **Note**: Works only when a `letterTag` is also specified
- `backwards` - removes letters one by one, by default `false` - removes the whole word
- `frugal` - find a common part during (backwards) clearing letters and start typing the new word from the queue if both words begin with exact same sentences. By default `false`. **Note**: Works only if `backwards` mode is enabled - set to `true`
- `leaveLast` - leave the last word visible after finishing the whole typing loop, by default `false`
- `infinity` - enable infinity type loop, by default `false`. **Note**: Option will be disabled automatically when both `infinity` and `leaveLast` options are selected
- `onType`/`onClear`/`onComplete` - functions which are triggered during plugin lifecycle. Accepts callback function with two parameters (`event` and `word`) as a value, by default `false`
### Events
We can distinguish three main events during the plugin lifecycle:
1. `onType` - triggered every time when the word is typing
2. `onClear` - triggered every time when the word is clearing
3. `onComplete` - triggered when typing loop is over and `infinity` option is disabled

## Test coverage
There are available some basic unit tests which (more or less) cover some basic *typeit.js* functionality.
Run `$yarn test` inside the project's root directory to start tests.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.0] - 2019-05-24
### Added
- `eslint` package to project

### Changed
- Move default typeit config values directly inside constructor
- Remove sort function from core and tests

### Fixed
- Tests plugin initialization because of module type changes

### Removed
- gulpfile.js file from project

## [1.2.0] - 2019-01-20
### Added
- `webpack` with simple config file

### Changed
- Remove `gulp` with related packages
- Plugin expose - now delivered as `CCS` lib under all the module definitions - `umd`
- Project directory structure - `/dist` folder contains only one production ready file

## [1.1.0] - 2018-11-17
### Added
- Frugal type mode
- Sorting function with test coverage

### Changed
- Babel preset used in a gulpfile to `babel-preset-env`
- Way of passing word in `initTypeLoop` - assign-by-value
- Array based words are transformed to strings on init (chunk mode was abandoned)

## [1.0.0] - 2018-09-14
### Added
- First release
- This CHANGELOG file
- Compiled dist js files


typeit.test.js
const typeit  = require('./dist/typeit')
const element = document.createElement('DIV')

element.setAttribute('id', 'sample');

describe('[ Functions ]', () => {
  let component

  beforeEach(function () {
    component = new typeit.TypeIt(element, {
      words: [
        'Sample text',
        'alpha',
        'Charlie',
        'yankee',
        'whiskey'
      ]
    })
  })

  test('[prepareLetter] Without specific configuration - simple text', () => {
    expect(component.prepareLetter('A')).toEqual('A')
  })

  test('[prepareLetter] With tag and class name configuration - html element', () => {
    component.letterTag = 'span'
    expect(component.prepareLetter('A')).toEqual(`<span class="">A</span>`)
  })

  test('[prepareLetter] With tag and class name configuration - html element', () => {
    component.letterTag   = 'span',
    component.letterClass = 'test'
    expect(component.prepareLetter('A')).toEqual(`<span class="test">A</span>`)
  })
})

describe('[ Promises ]', () => {
  let component

  beforeEach(function () {
    component = new typeit.TypeIt(element, {
      words: [
        'Sample text',
        'Sample code'
      ]
    })
  })

  it('[Wait] Delay 1500ms', () => {
    const start = performance.now()

    return component.wait(1500).then(data => {
      const end     = performance.now()
      const elapsed = end - start

      expect(elapsed).toBeGreaterThanOrEqual(data)
      expect(elapsed).toBeLessThanOrEqual(data + 150)
    })
  })

  it('[Clear] Word - both leaveLast and isLastWord are true', () => {
    component.leaveLast  = true
    component.isLastWord = true
    return component.clear('Sample').then(data => expect(data).toEqual('Sample'))
  })

  it('[Clear] Word - both leaveLast and isLastWord are false', () => {
    component.leaveLast  = false
    component.isLastWord = false
    return component.clear('Sample').then(data => expect(data).toEqual(''))
  })
})


gulpfile.js
var gulp         = require('gulp');
var autoprefixer = require('autoprefixer');
var postcss      = require('gulp-postcss');
var sass         = require('gulp-sass');
var babel        = require('gulp-babel');
var plumber      = require('gulp-plumber');
var eslint       = require('gulp-eslint');
var uglify       = require('gulp-uglify');
var rename       = require('gulp-rename');

// Asset paths
var paths = {
  sass: './sass/**/*.scss',
  js: './js/**/*.js',
  distJs: './dist/**/*.js'
};

gulp.task('default', ['sass', 'es6', 'watch']);

/**
 * Minify JavaScript
 */
gulp.task('compress', function () {
  return gulp.src(paths.distJs)
    .pipe(uglify())
    .pipe(rename('typeit.min.js'))
    .pipe(gulp.dest('./dist/js'))
});

/**
 * Compile Sass
 */
gulp.task('sass', function () {
  return gulp.src(paths.sass)
    .pipe(sass.sync().on('error', sass.logError))
    .pipe(postcss([
      autoprefixer({
        browsers: ['last 2 version']
      })
    ]))
    .pipe(gulp.dest('./dist/css'));
});

/**
 * Compile JS
 */
gulp.task('es6', function () {
  return gulp.src([
      'node_modules/babel-polyfill/dist/polyfill.js',
      paths.js
    ])
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(plumber())
    .pipe(babel({
      presets: ['babel-preset-env']
    }))
    .pipe(gulp.dest('./dist/js'));
});

/**
 * Watcher
 */
gulp.task('watch', function () {
  gulp.watch(paths.js, ['es6']);
  gulp.watch(paths.sass, ['sass']);
});


src/index.js
export class TypeIt {
  constructor(selector, config) {
    if (!config || !config.words) return

    this.backwards   = config.backwards || false
    this.delay       = isNaN(config.delay) ? 650 : Number(config.delay)
    this.element     = selector instanceof Element ? selector : document.querySelector(selector)
    this.events      = { onType: false, onClear: false, onComplete: false }
    this.frugal      = config.frugal || false
    this.initWord    = this.element.textContent
    this.isInfinity  = config.infinity && !config.leaveLast || false
    this.leaveLast   = config.leaveLast || false
    this.letterClass = config.letterClass || false
    this.letterTag   = config.letterTag || false
    this.onClear     = config.onClear || false
    this.onComplete  = config.onComplete || false
    this.onType      = config.onType || false
    this.speed       = isNaN(config.speed) ? 95 : Number(config.speed)
    this.words       = (this.initWord ? [this.initWord].concat(config.words) : config.words).map(
      elem => Array.isArray(elem) ? elem.join(' ') : elem
    )

    this.initBaseEvents()
    this.initTypeLoop()

    if (config.infinity && config.leaveLast) {
      console.warn('Both infinity and leaveLast options are selected. Loop will be disabled. Please update your settings.')
    }
  }

  initBaseEvents() {
    for (let eventName in this.events) {
      this.events[eventName] = new Event(eventName)
      this.element.addEventListener(eventName, (event) => {
        if (this[eventName] instanceof Function) {
          this[eventName].call(event, event, event.detail)
        }
      })
    }
  }

  updateBaseEventsDetailData(word) {
    for (let eventName in this.events) {
      this.events[eventName]['detail'] = word
    }
  }

  initTypeLoop() {
    for (
      let idx         = 0,
          typePromise = Promise.resolve(),
          currentWord = '';
      idx < this.words.length;
      idx++
    ) {
      // create shallow copy of current word
      currentWord = this.words[idx].slice()
      typePromise = typePromise.then(() => {
        this.isLastWord = this.words.length - 1 === idx
        this.updateBaseEventsDetailData(currentWord)

        return this.type(currentWord)
          .then(() => this.wait(this.delay))
          .then(() => this.clear(currentWord))
      }).then(() => {
        if (this.isInfinity && this.isLastWord) {
          this.initTypeLoop()
        }
        else if (!this.isInfinity && this.isLastWord) {
          this.element.dispatchEvent(this.events.onComplete)
        }
      })
    }
  }

  prepareLetter(letter) {
    return this.letterTag ?
      `<${this.letterTag} class="${this.letterClass || ''}">${letter}</${this.letterTag}>`
      :
      letter
  }

  type(word) {
    return new Promise(resolve => {
      if (
        this.frugal &&
        this.element.textContent &&
        this.isFrugalAvailable(word)
      ) {
        word = word.replace(this.element.textContent, '')
      }
      else if (this.element.textContent) {
        resolve(word)
        return
      }

      this.element.dispatchEvent(this.events.onType)

      let cursorPosition = 0
      const typeLoop = setInterval(() => {
        if (word[cursorPosition]) {
          this.element.insertAdjacentHTML(
            'beforeend',
            this.prepareLetter(word[cursorPosition])
          )
        }
        else {
          clearInterval(typeLoop)
          resolve(word)
          return
        }

        cursorPosition++
      }, this.speed)
    })
  }

  wait(time) {
    return new Promise(resolve => setTimeout(() => resolve(time), time))
  }

  clear(word) {
    return new Promise(resolve => {
      if (this.leaveLast && this.isLastWord) {
        resolve(word)
        return
      }

      let currentWordIndex = this.words.indexOf(word)

      this.element.dispatchEvent(this.events.onClear)

      const clearLoop = setInterval(() => {
        if (this.backwards && this.element.textContent) {
          this.element.removeChild(this.element.lastChild)
          if (
            this.frugal &&
            this.isFrugalAvailable(
              this.isLastWord &&
              this.isInfinity ? this.words[0] : this.words[currentWordIndex + 1]
            )
          ) {
            resolve(word)
            clearInterval(clearLoop)
          }
        }
        else {
          this.element.textContent = ''
          clearInterval(clearLoop)
          resolve(this.element.textContent)
          return
        }
      }, this.speed)
    })
  }

  isFrugalAvailable(a) {
    return new RegExp(`^(${this.element.textContent}).*`).test(a)
  }
}


js/typeit.js
const DEFAULT_DELAY        = 650
const DEFAULT_SPEED        = 95
const DEFAULT_TYPEIT_CLASS = 'typeit'

class Typeit {
  constructor(selector, config) {
    if (!config || !config.words) return

    this.backwards   = config.backwards || false
    this.delay       = config.delay || DEFAULT_DELAY
    this.element     = selector instanceof Element ? selector : document.querySelector(selector)
    this.events      = { onType: false, onClear: false, onComplete: false }
    this.frugal      = config.frugal || false
    this.initWord    = this.element.textContent
    this.leaveLast   = config.leaveLast || false
    this.isInfinity  = config.infinity && !config.leaveLast || false
    this.letterClass = config.letterClass || false
    this.letterTag   = config.letterTag || false
    this.onClear     = config.onClear || false
    this.onComplete  = config.onComplete || false
    this.onType      = config.onType || false
    this.sort        = config.sort || false
    this.speed       = config.speed || DEFAULT_SPEED
    this.words       = (this.initWord ? [this.initWord].concat(config.words) : config.words)
      .map(elem => Array.isArray(elem) ? elem.join(' ') : elem)

    this.element.classList.add(DEFAULT_TYPEIT_CLASS)
    if (this.sort) this.words = this.sortWords()

    this.initBaseEvents()
    this.initTypeLoop()

    if (config.infinity && config.leaveLast) {
      console.warn('Both infinity and leaveLast options are selected. Loop will be disabled. Please update your settings.')
    }
  }

  initBaseEvents() {
    for (let eventName in this.events) {
      this.events[eventName] = new Event(eventName)
      this.element.addEventListener(eventName, (event) => {
        if (this[eventName] instanceof Function) {
          this[eventName].call(event, event, event.detail)
        }
      })
    }
  }

  updateBaseEventsDetailData(word) {
    for (let eventName in this.events) {
      this.events[eventName]['detail'] = word
    }
  }

  initTypeLoop() {
    for (
      let idx         = 0,
          typePromise = Promise.resolve(),
          currentWord = '';
      idx < this.words.length;
      idx++
    ) {
      // create shallow copy of current word
      currentWord = this.words[idx].slice()
      typePromise = typePromise.then(() => {
        this.isLastWord = this.words.length - 1 === idx
        this.updateBaseEventsDetailData(currentWord)

        return this.type(currentWord)
          .then(() => this.wait(this.delay))
          .then(() => this.clear(currentWord))
      }).then(() => {
        if (this.isInfinity && this.isLastWord) {
          this.initTypeLoop()
        }
        else if (!this.isInfinity && this.isLastWord) {
          this.element.dispatchEvent(this.events.onComplete)
        }
      })
    }
  }

  prepareLetter(letter) {
    return this.letterTag ?
      `<${this.letterTag} class="${this.letterClass || ''}">${letter}</${this.letterTag}>`
      :
      letter
  }

  type(word) {
    return new Promise(resolve => {
      if (
        this.frugal &&
        this.element.textContent &&
        this.isFrugalAvailable(word)
      ) {
        word = word.replace(this.element.textContent, '')
      }
      else if (this.element.textContent) {
        resolve(word)
        return
      }

      this.element.dispatchEvent(this.events.onType)

      let cursorPosition = 0
      const typeLoop = setInterval(() => {
        if (word[cursorPosition]) {
          this.element.insertAdjacentHTML(
            'beforeend',
            this.prepareLetter(word[cursorPosition])
          )
        }
        else {
          clearInterval(typeLoop)
          resolve(word)
          return
        }

        cursorPosition++
      }, this.speed)
    })
  }

  wait(time) {
    return new Promise(resolve => setTimeout(() => resolve(time), time))
  }

  clear(word) {
    return new Promise(resolve => {
      if (this.leaveLast && this.isLastWord) {
        resolve(word)
        return
      }

      let currentWordIndex = this.words.indexOf(word)

      this.element.dispatchEvent(this.events.onClear)

      const clearLoop = setInterval(() => {
        if (this.backwards && this.element.textContent) {
          this.element.removeChild(this.element.lastChild)
          if (
            this.frugal &&
            this.isFrugalAvailable(
              this.isLastWord &&
              this.isInfinity ? this.words[0] : this.words[currentWordIndex + 1]
            )
          ) {
            resolve(word)
            clearInterval(clearLoop)
          }
        }
        else {
          this.element.textContent = ''
          clearInterval(clearLoop)
          resolve(this.element.textContent)
          return
        }
      }, this.speed)
    })
  }

  isFrugalAvailable(a) {
    return new RegExp(`^(${this.element.textContent}).*`).test(a)
  }

  sortWords() {
    return this.words.sort((a, b) => {
      a = a.toLowerCase()
      b = b.toLowerCase()
      return a === b ? 0 : a < b ? -1 : 1
    })
  }
}

try {
  module.exports = Typeit
}
catch (error) {
  console.warn(error)
}
